features/26-mfa/api.md

```diff
@@ -176,10 +177,11 @@ try {
 // Present AN UI to let the user to optionally name the TOTP authenticator.
 
 const displayName = textInput.value;
-const { authenticatorID, secret } = await skygear.auth.mfa.createNewTOTP(displayName);
+const { authenticatorID, secret, otpauthURI } = await skygear.auth.mfa.createNewTOTP(displayName);
 
 // Present to the secret to the user and let the user
 // to add the secret to their TOTP application, such as Authy and Google Authenticator.
+// Or preferably use a OR code to display otpauthURI, it is compatible with Google Authenticator and Authy.
```

Comment:
I don't think we should provide a otpauthURI; the app should generate it at client side with the provided secret, since most of the parameters in URI seems app-specific (e.g. issuer, account name).

Also, typo: OR should be QR?

Reply:
SDK is client-side, and for usability we definitely need otpauthURI. In fact, I think on Client-side, we should have generated the QR code for users to directly display it, instead of "generate QR code from the otpauthURI".

(and it should depend on the platform, so on Web we probably need server-side to generate the QR Code and return the URL in the form of a URL; On iOS / Android we can generate the QR code on client-side devices and return it as some sort of object that users can use to display easily)

Reply:
> since most of the parameters in URI seems app-specific (e.g. issuer, account name).

I believe the parameters should be relatively static so they can be configurable in tenant config.

> so on Web we probably need server-side to generate the QR Code and return the URL in the form of a URL

Good suggestion! It should be possible to add an endpoint to serve a QR code image. Let me include this.

Reply:
> I believe the parameters should be relatively static so they can be configurable in tenant config.

Account name can be customized, e.g. user display name from self-managed database, role of user, etc.

Reply:
The specification of otpauth is here https://github.com/google/google-authenticator/wiki/Key-Uri-Format

The only configurable input to generate otpauth URI is "Issuer" and "Account Name". Since a user can have many login IDs, the Auth gear cannot find a good candidate for "Account Name".

I think instead of generating the otpauth URI on server-side, the client SDK should offer an utility function to generate otpauth URI.

To generate an QR code image, the developer should first use our client SDK to generate otpauth URI. The server offers an endpoint to accept a otpauth URI and return an QR code image.

Reply:
@louischan-oursky the specification didn't specify what is account name, only say it is for identifying the account uniquely.

I'm a bit confused, if Auth gear cannot find a good candidate for Account Name, how would Client SDK able to choose it? (or how Developers are supposed to choose it?)

Reply:
Both the "Issuer" and "Account Name" is for display purpose in the Authenticator Application such as Authy and Google Authenticator. Only the secret is used in deriving the Time-based one-time password. The primary purpose of issuer and account name is for the user to identify registered secrets. If the application uses email as login ID, then most likely the developer should use the email of the user as "Account Name". Similarly "Issuer" should be a user-facing application name chosen by the developer.

It is perfectly fine to have duplicate "Issuer" and "Account Name" pair as illustrated by the following screenshot.

![IMG_0185](https://user-images.githubusercontent.com/15082564/61617759-f6aebb00-ac9d-11e9-90c0-8d77ce5c47e3.PNG)

Reply:
ic, sounds like we should let user define their own issuer and account name.

I'm thinking if there is any security implication... would some how a social attack could be engineered to make users fake a OTP with a phishing website' issuer and account name? hmmm

Reply:
> I'm thinking if there is any security implication for client side, would some how a social attack could be engineered to make users fake a OTP with a phishing website' issuer and account name? hmmm

The one-time password is short (6 digits) and valid for 30+-30s (1.5 min). So TOTP is subject to brute force attack if we do not have authentication attempt limits. We can discuss authentication attempt limits in another issue.

The details of security of TOTP (which is an extension of HOTP) can be found in the RFC

- https://tools.ietf.org/html/rfc4226#appendix-A.1
- https://tools.ietf.org/html/rfc4226#appendix-B.1

Reply:
Oh sure... but it seems very important and we should mention the limits in this spec and create issues for repo?

Reply:
I'll mention the limits in this spec and open a feature issue for authentication attempt limits.

Reply:
Issue opened here https://github.com/SkygearIO/features/issues/344